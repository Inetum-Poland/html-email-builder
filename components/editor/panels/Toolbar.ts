import type { Editor, PanelProperties } from "grapesjs";
import { Scan, FileCode2, Undo, Redo, FileJson2 } from "lucide-static";
import { CodePreview, exportProject, Format } from "./components/CodePreview";

export const Toolbar = (): PanelProperties => {
  const { $i18n: { t } } = useNuxtApp();

  return {
    id: "toolbar",
    buttons: [
      {
        attributes: { title: t("exportHTML"), class: "button-group-left" },
        command: (editor: Editor) => {
          const html = editor.Commands.run("mjml-code-to-html").html;
          const mjml = editor.Commands.run("mjml-code");

          const wrapper = document.createElement("div");
          wrapper.className = "code-preview";
          wrapper.append(CodePreview(editor, Format.MJML, mjml));
          wrapper.append(CodePreview(editor, Format.HTML, html));

          editor.Modal.setTitle(t("exportCode")).setContent(wrapper).open();
        },
        id: "export-html-modal",
        label: FileCode2,
        tagName: "button",
      },
      {
        attributes: {
          title: t("copyFormattedDocument"),
          id: "copy-document-button",
          class: "button-group-right",
        },
        command: async (editor: Editor) => {
          const button = document.querySelector("#copy-document-button");

          if (!button) {
            return;
          }

          const data = editor.Commands.run("mjml-code-to-html").html;
          const html = new Blob([data], { type: "text/html" });
          const content = new ClipboardItem({ "text/html": html });
          await navigator.clipboard.write([content]);
          button.innerHTML = t("copied");
          setTimeout(() => (button.innerHTML = t("copyDocument")), 3000);
        },
        id: "export-html",
        label: t("copyDocument"),
        tagName: "button",
      },
      {
        attributes: { title: t("exportJSON"), class: "button-group-left" },
        command: (editor: Editor) => {
          const json = JSON.stringify(editor.getProjectData(), null, 2);

          const wrapper = document.createElement("div");
          wrapper.className = "code-preview";
          wrapper.append(CodePreview(editor, Format.JSON, json));

          editor.Modal.setTitle(t("exportProjectState")).setContent(wrapper).open();
        },
        id: "export-json-modal",
        label: FileJson2,
        tagName: "button",
      },
      {
        attributes: {
          title: t("exportProjectState"),
          class: "button-group-right",
          id: "export-project-button",
        },
        command: (editor: Editor) => {
          const json = JSON.stringify(editor.getProjectData(), null, 2);
          exportProject(Format.JSON, json);
        },
        id: "export-json",
        label: t("saveProject"),
        tagName: "button",
      },
      {
        attributes: { title: t("toggleBorders") },
        command: "sw-visibility",
        id: "toggle-borders",
        label: Scan,
        tagName: "button",
      },
      {
        attributes: { title: t("undo") },
        command: "core:undo",
        id: "undo",
        label: Undo,
        tagName: "button",
      },
      {
        attributes: { title: t("redo") },
        command: "core:redo",
        id: "redo",
        label: Redo,
        tagName: "button",
      },
    ],
  };
};
