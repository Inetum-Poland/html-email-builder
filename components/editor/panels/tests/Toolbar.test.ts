import { expect, test } from "vitest";
import { Toolbar } from "../Toolbar";

test("renders the Toolbar panel", () => {
  expect(Toolbar()).toMatchSnapshot();
});
