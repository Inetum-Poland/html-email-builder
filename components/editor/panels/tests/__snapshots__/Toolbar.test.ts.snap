// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`renders the Toolbar panel 1`] = `
{
  "buttons": [
    {
      "attributes": {
        "class": "button-group-left",
        "title": "Export HTML",
      },
      "command": [Function],
      "id": "export-html-modal",
      "label": "
<svg
  class="lucide lucide-file-code-2"
  xmlns="http://www.w3.org/2000/svg"
  width="24"
  height="24"
  viewBox="0 0 24 24"
  fill="none"
  stroke="currentColor"
  stroke-width="2"
  stroke-linecap="round"
  stroke-linejoin="round"
>
  <path d="M4 22h14a2 2 0 0 0 2-2V7l-5-5H6a2 2 0 0 0-2 2v4" />
  <path d="M14 2v4a2 2 0 0 0 2 2h4" />
  <path d="m5 12-3 3 3 3" />
  <path d="m9 18 3-3-3-3" />
</svg>
",
      "tagName": "button",
    },
    {
      "attributes": {
        "class": "button-group-right",
        "id": "copy-document-button",
        "title": "Copy formatted document",
      },
      "command": [Function],
      "id": "export-html",
      "label": "Copy document",
      "tagName": "button",
    },
    {
      "attributes": {
        "class": "button-group-left",
        "title": "Export JSON",
      },
      "command": [Function],
      "id": "export-json-modal",
      "label": "
<svg
  class="lucide lucide-file-json-2"
  xmlns="http://www.w3.org/2000/svg"
  width="24"
  height="24"
  viewBox="0 0 24 24"
  fill="none"
  stroke="currentColor"
  stroke-width="2"
  stroke-linecap="round"
  stroke-linejoin="round"
>
  <path d="M4 22h14a2 2 0 0 0 2-2V7l-5-5H6a2 2 0 0 0-2 2v4" />
  <path d="M14 2v4a2 2 0 0 0 2 2h4" />
  <path d="M4 12a1 1 0 0 0-1 1v1a1 1 0 0 1-1 1 1 1 0 0 1 1 1v1a1 1 0 0 0 1 1" />
  <path d="M8 18a1 1 0 0 0 1-1v-1a1 1 0 0 1 1-1 1 1 0 0 1-1-1v-1a1 1 0 0 0-1-1" />
</svg>
",
      "tagName": "button",
    },
    {
      "attributes": {
        "class": "button-group-right",
        "id": "export-project-button",
        "title": "Export project state",
      },
      "command": [Function],
      "id": "export-json",
      "label": "Save project",
      "tagName": "button",
    },
    {
      "attributes": {
        "title": "Toggle borders",
      },
      "command": "sw-visibility",
      "id": "toggle-borders",
      "label": "
<svg
  class="lucide lucide-scan"
  xmlns="http://www.w3.org/2000/svg"
  width="24"
  height="24"
  viewBox="0 0 24 24"
  fill="none"
  stroke="currentColor"
  stroke-width="2"
  stroke-linecap="round"
  stroke-linejoin="round"
>
  <path d="M3 7V5a2 2 0 0 1 2-2h2" />
  <path d="M17 3h2a2 2 0 0 1 2 2v2" />
  <path d="M21 17v2a2 2 0 0 1-2 2h-2" />
  <path d="M7 21H5a2 2 0 0 1-2-2v-2" />
</svg>
",
      "tagName": "button",
    },
    {
      "attributes": {
        "title": "Undo",
      },
      "command": "core:undo",
      "id": "undo",
      "label": "
<svg
  class="lucide lucide-undo"
  xmlns="http://www.w3.org/2000/svg"
  width="24"
  height="24"
  viewBox="0 0 24 24"
  fill="none"
  stroke="currentColor"
  stroke-width="2"
  stroke-linecap="round"
  stroke-linejoin="round"
>
  <path d="M3 7v6h6" />
  <path d="M21 17a9 9 0 0 0-9-9 9 9 0 0 0-6 2.3L3 13" />
</svg>
",
      "tagName": "button",
    },
    {
      "attributes": {
        "title": "Redo",
      },
      "command": "core:redo",
      "id": "redo",
      "label": "
<svg
  class="lucide lucide-redo"
  xmlns="http://www.w3.org/2000/svg"
  width="24"
  height="24"
  viewBox="0 0 24 24"
  fill="none"
  stroke="currentColor"
  stroke-width="2"
  stroke-linecap="round"
  stroke-linejoin="round"
>
  <path d="M21 7v6h-6" />
  <path d="M3 17a9 9 0 0 1 9-9 9 9 0 0 1 6 2.3l3 2.7" />
</svg>
",
      "tagName": "button",
    },
  ],
  "id": "toolbar",
}
`;
